packages/web-react/src/hooks/useStyleUtilities.ts

```diff
@@ -0,0 +1,51 @@
+import { SpacingStyleProp } from '../constants';
+import { BreakpointToken, SpaceToken, StyleProps } from '../types';
+import { applyClassNamePrefix } from '../utils';
+
+export type StyleUtilitiesResult = {
+  styleUtilities: string[];
+  props: StyleProps;
+};
+
+const normalizeSpacingValue = (value: string): string => (value === 'auto' ? 'auto' : value.replace(/[^0-9]/g, ''));
+
+const processBreakpointProperties = (
+  utilityName: string,
+  propValue: Partial<Record<BreakpointToken, SpaceToken | 'auto'>>,
+  prefix: string | null | undefined,
+  accumulatedUtilities: string[] = [],
+) => {
+  Object.keys(propValue).forEach((breakpoint) => {
+    const breakpointValue = propValue[breakpoint as keyof typeof propValue];
+    const utilityValue = normalizeSpacingValue(breakpointValue as string);
+    const infix = breakpoint === 'mobile' ? '' : `-${breakpoint}`;
+    accumulatedUtilities.push(applyClassNamePrefix(prefix)(`${utilityName}${infix}-${utilityValue}`));
+  });
+
+  return accumulatedUtilities;
+};
+export function useStyleUtilities(props: StyleProps, prefix: string | null | undefined): StyleUtilitiesResult {
+  const styleUtilities = Object.keys(props).reduce((accumulatedUtilities: string[], key) => {
+    const propValue = props[key as keyof StyleProps] as Partial<Record<BreakpointToken, SpaceToken | 'auto'>>;
+
+    if (Object.keys(SpacingStyleProp).includes(key)) {
+      const utilityName = SpacingStyleProp[key as keyof typeof SpacingStyleProp];
+
+      if (typeof propValue === 'object' && propValue !== null) {
+        processBreakpointProperties(utilityName, propValue, prefix, accumulatedUtilities);
+      } else if (propValue) {
+        const utilityValue = normalizeSpacingValue(propValue as string);
+        accumulatedUtilities.push(applyClassNamePrefix(prefix)(`${utilityName}-${utilityValue}`));
+      }
+
+      delete props[key as keyof typeof props];
```

Comment:
I believe it should be possible (and preferable) to work with the data using the array methods `map()`, `filter()`, or `reduce()`. As far as good my knowledge of JS is (and you know how good it is), modifying an outside object with `push()`, `delete()` etc. is an undesired design pattern. It is better to use a more functional approach, not only because such code can be tested more thoroughly and does not have any side effects.

Reply:
refactored